### PR TITLE
Fix gcc8 compilation error

### PR DESCRIPTION
**Pull request by avramidis**
This fix is for issue #4419. The solution seems to work for g++ (Ubuntu 8.2.0-7ubuntu1) 8.2.0 but it might need some tinkering.

### CODE REVIEW

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 1]
```diff
@@ -764,20 +790,6 @@ namespace shogun
...
-/*
- * Eigen's symmetric eigensolver uses a slower algorithm in comparison
- * to LAPACK's dsyevr, so if LAPACK is available we use it for float64 type.
- * This should be removed if eventually Eigen will provide a faster
- * symmetric eigensolver (@see
- * http://eigen.tuxfamily.org/bz/show_bug.cgi?id=522).
- */
-#ifdef HAVE_LAPACK
-	template <>
-	void LinalgBackendEigen::eigen_solver_symmetric_impl<float64_t>(
```
Comment (gf712) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: But now there is no float64_t specialisation? Or is there now a faster symmetric eigensolver provided by Eigen?

**Comment by gf712** [comment]
Once we reach C++17 this will be much simpler as we can just dispatch with `if constexpr(...)`. Until then I think you need to do SFINAE, since gcc8 doesn't like this anymore. You shouldn't do partial specialisations of class members.
It's not ideal but you could go around it by adding a `std::enable_if_t<std::is_same<T, float64_t>::value>* = nullptr` and `std::enable_if_t<!std::is_same<T, float64_t>::value>* = nullptr` for each case and then it should compile fine.

**Comment by avramidis** [comment]
Thanks! I will try this solution.

**Comment by gf712** [comment]
Nevermind, you end up with the same linker error. I am not sure how to fix it... you might have to just move the stuff from the translation unit to the header?

**Comment by avramidis** [comment]
OK. I need to build shogun on windows somehow and try to fix the problem there.

**Comment by avramidis** [comment]
OK, seems to work fine now. With these changes I can build with gcc8 in Ubuntu 18.10.

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 2]
```diff
@@ -557,8 +558,66 @@ namespace shogun
...
+			eigenvalues_eig = solver.eigenvalues().tail(k).template cast<T>();
+			eigenvectors_eig = solver.eigenvectors().rightCols(k).template cast<T>();
+		}
+
+		/*
+		* Eigen's symmetric eigensolver uses a slower algorithm in comparison
+		* to LAPACK's dsyevr, so if LAPACK is available we use it for float64 type.
+		* This should be removed if eventually Eigen will provide a faster
+		* symmetric eigensolver (@see
+		* http://eigen.tuxfamily.org/bz/show_bug.cgi?id=522).
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: did you check whether this eigen issue was solved recently?

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 3]
```diff
@@ -557,8 +558,66 @@ namespace shogun
...
+		*/
+		#ifdef HAVE_LAPACK
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: mmh does this make sense to guard this here?

Comment (karlnapf) [reply src/shogun/mathematics/linalg/LinalgBackendEigen.h]: ah nevermind

**Comment by avramidis** [comment]
Will do!

**Comment by karlnapf** [comment]
Cool, and?

**Comment by avramidis** [comment]
@karlnapf sorry I was working on another project for the last few days. After work today I started working on my PRs again and hopefully this weekend with finish them :)

**Comment by avramidis** [comment]
Benchmarks results with #4519 

**Before gcc8 fix  with gcc7**

Start 1: RandomFourierDotFeatures_benchmark
1/6 Test 1: RandomFourierDotFeatures_benchmark ...   Passed   22.11 sec
    Start 2: HashedDocDotFeatures_benchmark
2/6 Test 2: HashedDocDotFeatures_benchmark .......   Passed   13.90 sec
    Start 3: RefCount_benchmark
3/6 Test 3: RefCount_benchmark ...................   Passed    0.69 sec
    Start 4: BasicOps_benchmark
4/6 Test 4: BasicOps_benchmark ...................   Passed   37.75 sec
    Start 5: Misc_benchmark
5/6 Test 5: Misc_benchmark .......................   Passed   35.13 sec
    Start 6: SGMatrix_benchmark
6/6 Test 6: SGMatrix_benchmark ...................   Passed   17.60 sec

100% tests passed, 0 tests failed out of 6

Label Time Summary:
benchmark    = 127.19 sec*proc (6 tests)

Total Test time (real) = 127.20 sec

**After gcc8 fix with gcc7**

Start 1: RandomFourierDotFeatures_benchmark
1/6 Test 1: RandomFourierDotFeatures_benchmark ...   Passed   21.54 sec
    Start 2: HashedDocDotFeatures_benchmark
2/6 Test 2: HashedDocDotFeatures_benchmark .......   Passed   13.65 sec
    Start 3: RefCount_benchmark
3/6 Test 3: RefCount_benchmark ...................   Passed    0.69 sec
    Start 4: BasicOps_benchmark
4/6 Test 4: BasicOps_benchmark ...................   Passed   37.67 sec
    Start 5: Misc_benchmark
5/6 Test 5: Misc_benchmark .......................   Passed   34.92 sec
    Start 6: SGMatrix_benchmark
6/6 Test 6: SGMatrix_benchmark ...................   Passed   16.87 sec

100% tests passed, 0 tests failed out of 6

Label Time Summary:
benchmark    = 125.34 sec*proc (6 tests)

Total Test time (real) = 125.35 sec

**Comment by karlnapf** [comment]
ah just saw this nevermind my other comment in #4519

**Comment by karlnapf** [comment]
sweet so no changes, sounds good thanks for doing this!

**Pull request merged**